### ISSUE TITLE
Any plans to support substitutionGroup ?

### ISSUE THREAD

**Issue by Liwoj** [mganss/XmlSchemaClassGenerator#36]
https://msdn.microsoft.com/en-us/library/awac9czf(v=VS.100).aspx#SubstitutionGroup Attribute

**Comment by mganss** [comment]
AFAICT what's missing is support for the case where the head is abstract. I won't have time to implement this currently, perhaps in a few weeks. If you could implement this, that'd be great 😄

**Comment by Liwoj** [comment]
Yep, abstract head is exactly what i was doing. Thank for clarification. 
I don't have time to implement it now either

**Comment by John-Welch** [comment]
I have a partial implementation aimed at solving a specific case I use frequently in my code. I want an xml file such as
```xml
<?xml version="1.0" encoding="utf-8"?>
<fleet xmlns="http://schemas.example.com/test">
  <car2 name="Beetle"/>
  <bus2 name="Routemaster"/>
  <car2 name="Mini"/>
</fleet>
```
to read into classes like
```C#
abstract class Vehicle
{
    public string Name {get; set;}
}
class Car: Vehicle { }
class Bus: Vehicle { }

class Fleet
{
    public List<Vehicle> Vehicle {get; set;}
}
```
so I define an xml schema as follows:
```xml
<?xml version="1.0" encoding="utf-8"?>
<schema targetNamespace="http://schemas.example.com/test"
    elementFormDefault="qualified"
    xmlns="http://www.w3.org/2001/XMLSchema"
    xmlns:test="http://schemas.example.com/test"
>
  <complexType name="vehicle" abstract="true">
    <attribute name="name" type="string"/>
  </complexType>
  <complexType name="car" final="#all">
    <complexContent>
      <extension base="test:vehicle"/>
    </complexContent>
  </complexType>
  <complexType name="bus" final="#all">
    <complexContent>
      <extension base="test:vehicle"/>
    </complexContent>
  </complexType>

  <element name="vehicle" type="test:vehicle" substitutionGroup="test:vehicle" abstract="true"/>
  <element name="car2" type="test:car" substitutionGroup="test:vehicle"/>
  <element name="bus2" type="test:bus" substitutionGroup="test:vehicle"/>
  
  <element name="fleet">
    <complexType>
      <sequence>
        <element maxOccurs="unbounded" ref="test:vehicle"/>
      </sequence>
    </complexType>
  </element>
</schema>
```
This generates almost-working C# classes with xscgen v2.0.160; it is just missing `        [System.Xml.Serialization.XmlElementAttribute("car2", Type=typeof(Car), Namespace="http://schemas.example.com/test")]` (and similar for Bus) attributes before the `List<Vehicle>` property definition.

I've got a PR where it uses the `car2` etc. root elements' `substitutionGroup` attribute to identify that `vehicle` is substitutable, and remembers the `car2` substitution name. I adjusted the existing code generator to use the substitution name if it is available when emitting the attributes.

There are a couple of known limitations with this:
1. The `substitutionGroup` text value is still ignored as in the existing code, so xscgen will emit attributes for all its descendent classes if the head is in a substitution group, irrespective of whether those descendents are also in the substitution group.
2. The code doesn't fully distinguish between the `<complexType>` definitions and the root `<element>`s that form the substitution group. For instance, if a `<element name="car3" type="test:car" substitutionGroup="test:vehicle"/>` were added to the above schema, only one attribute would be emitted (for car3, as that would overwrite car2 because it comes last).

Neither of these limitations affect what I want to do and won't be trivial to fix, so I've left them.
